### PR TITLE
TGP-2703 Assessment page changes + related refactors

### PR DESCRIPTION
**Pull request by Ewan-James-Donovan**
# **PR README**

https://jira.tools.tax.service.gov.uk/browse/TGP-2703

## Overview
The ticket initially seemed to be a simple UI change but, for three reasons, this turned out not to be the case. Initially the ticket asked for:
-	some easy content changes
-	the yes/no radio buttons on assessment question pages to be changed to checkboxes for the exemptions *(Reason 1)*
-	to display the measure description to the user *(Reason 2)*

And then the ticket was updated following conversations with OTT so that now:
- the measure won't be displayed, but the theme will be instead *(Reason 3)*

*Reason 1* means that the answer type fundamentally changes from a boolean to a list or sequence.
*Reason 2* means that development was slowed and a fair bit of work was deleted, I've kept the code on a local branch in case we ever need it.
*Reason 3* means that we now need to parse and handle the theme's string description from OTT's response, which wasn't being done prior to this ticket.

So, in essence, this PR is 4 things:
1) UI changes
2) Changes to how the answer is stored and handled
3) Additions to OTT parsing regarding themes
4) Tests and test refactors as necessitated by the above

## Changes to each file

[app/controllers/AssessmentController.scala](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-5cc82d4c7ce7459357325e012803d0b232523a6bdd1ca7793c8c898e3be68a58):
- now passes `assessment.themeDescription` to the view.
- now passes `assessment.getCodesZippedWithDescriptions` down to the view instead of `listItems`, keeping the exemption codes and descriptions separate.
- passes total number of assessments to the view for the title to include

[app/controllers/BaseController.scala](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-5cc82d4c7ce7459357325e012803d0b232523a6bdd1ca7793c8c898e3be68a58):
- `prepareForm` method refactored to work when the Form and QuesstionPage types are different.

[app/forms/AssessmentFormProvider.scala](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-20b671ca0bfdc37633d0f9fa76f5afca9188faa2f79c2e63899fcb6b7a61a3d8):
- Refactored to accept a list of answers from the user’s checkbox input
- Calls `AssessmentAnswer.fromSeq / toSeq` to convert back and forth between the answer the user gave and our `AssesmentAnswer` model
- Now only uses one error message, all usages of the form apply method now do not have the exemption count.

[app/models/AnsweredQuestions.scala](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-dd47f1ccd1e68068f2e169abdd23c5c45e87a6d0e53a7b7857513eed41e2bcc3):
- `isAnswered` method now matches for Exemption containing list of checked answers

[app/models/AssessmentAnswer.scala](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-6055bcf6ee7cb9974762d0f58b3dd4130fb890fd71d98f3af1551be16d2b2689)
- `AssessmentAnswer.Exemption` Has been refactored to allow a list of checked answers
-  `writes` and `reads` implicit methods on the companion object have been refactored to be able to parse and serialise the list mentioned
- `fromString` has been replaced with `fromSeq(input: Seq[String])` to reflect the new input type.
- `toSeq` has been added as an inverse function of `fromSeq` to keep things clean.

[app/models/ott/CategoryAssessment.scala](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-b6f667bef861ef94bf3e3fbc6d5a3ea77f5ecb5c59d46ec00fc2efee354bcbeb)
- Case class now has a `themeDescription` property which holds the theme description related to the current category assessment.
- Now has a method called `getCodesZippedWithDescriptions` which zips together exemption codes and descriptions. It replaces `getExemptionListItems`, which previously concatenated both into a string. This is so only the code can be saved to user answers.
- The `build` from ottResponse method on the companion object now populates `themeDescription` from the included elements in the OTT response.

[app/models/ott/response/ThemeResponse.scala](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-04b1957cf4961aa5c04d586c78cf9386ff3e8c59573f796def4b9170f6b63f37)
- Case class now contains a property for `theme: String`, which holds a description of the theme from OTT.
- Companion object `reads` method now parses out `theme`.

[app/navigation/Navigator.scala](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-853d6712e99a36da2e88ad7fad3c0fb9014e43a9cb23ae690a7c99a02b242c6c)
- `Exemption` pattern matching has changed slightly since it accepts a list of answers.

[app/views/AssessmentView.scala.html](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-a5140e162c7a80cbaac617731cf53aed6785c7f1825b97db5f545003060e872e)
- Now accepts `codesAndDescriptions` instead of `listItems`
- Replaced radios with`govukCheckboxes`
- Now accepts and displays `themeDescription`
- Now accepts and displays total assessments in title
- Layout changes as per ticket

[conf/messages.en](https://github.com/hmrc/trader-goods-profiles-frontend/pull/523/files#diff-0e90ad12b779fdc21efc1c10d0c58baa68136b35bedc5a6a0593f8e9479db567)
- Messages changed as per ticket + welsh in the other welsh file, not all welsh was supplied.

There are many changes in the test directory, but essentially, they have been updated to include and assert for these measure type descriptions, as well as handle `Exemption` with a list of codes.

The form provider has some extra testing written to improve code coverage.

----

# Checklist for developers before reviewing
 - [x] I have merged in main
 - [x] I have assigned the PR to me
 - [x] I ran the code with a clean compile test and ensured there are no warnings or deprecations
 - [x] I have run `sbt prePR` to run all tests, format and check the code coverage
 - [x] I have written unit tests to cover any code that I wrote, where necessary
 - [x] I have linked tickets to any TODOs I have written
 - [x] I have run the microservices with the latest SM2 config
 - [x] I have checked all the Acceptance Criteria pass
 - [x] I have run the pipeline journey tests
 - [x] I have run the local journey tests
 - [x] I have spoken to the QAs about any changes that may need to be made to the journey tests OR no changes to the journey tests are necessary
 - [x] I have moved the ticket to Review in Jira
 - [x] I have left a message in tgp-prs on slack to alert other Devs that this ticket is ready for review
 - [x] I have checked my PR is not in draft mode

# Checklist for reviewer 1
- [x] I have added an emoji to the relevant message in tgp-prs to show I am reviewing
- [x] I have checked out the branch
- [x] I have run the microservices with the latest SM2 config
- [x] I have checked all the Acceptance Criteria pass
- [x] I reviewed the code and left comments where necessary
- [x] I have left a message on the relevant tgp-prs message to show that I have finished reviewing

# Checklist for reviewer 2
- [x] I have added an emoji to the relevant message in tgp-prs to show I am reviewing
- [x] I have checked out the branch
- [x] I have run the microservices with the latest SM2 config
- [ ] I have checked all the Acceptance Criteria pass
- [ ] I reviewed the code and left comments where necessary
- [ ] I have left a message on the relevant tgp-prs message to show that I have finished reviewing

# Checklist for developers before merging
 - [x] The PR has been approved by 2 other developers
 - [x] I have merged in main
 - [x] I have run `sbt preMerge` to run all tests, check formatting and code coverage
 - [x] I have run the microservices with the latest SM2 config
 - [x] I have checked all the Acceptance Criteria pass
 - [x] I have checked that the QAs have merged in any changes to the journey tests
 - [x] I have run the pipeline journey tests
 - [x] I have run the local journey tests

# Reminders for developers after merging
 1. Wait for pipeline to pass
 2. Move ticket to QA column in Jira
 3. Leave message in tgp-qa on slack to alert QAs that this ticket is ready for testing

### CODE REVIEW

File: app/forms/AssessmentFormProvider.scala [hunk 1]
```diff
@@ -19,22 +19,31 @@ package forms
...
+
     Form(
       "value" ->
-        text(messagesKey)
+        seq(text(messagesKey))
+          .verifying(messagesKey, values => values.nonEmpty)
           .verifying(
             messagesKey,
-            value => Set("true", "false").contains(value)
+            values => !values.contains("false") || values.size == 1
```
Comment (Ewan-James-Donovan) [inline app/forms/AssessmentFormProvider.scala:37]: can't remember why I did this, double check it

File: app/forms/AssessmentFormProvider.scala [hunk 2]
```diff
@@ -19,22 +19,27 @@ package forms
...
+
     Form(
       "value" ->
-        text(messagesKey)
+        seq(text(messagesKey))
+          .verifying(messagesKey, values => values.nonEmpty)
           .verifying(
             messagesKey,
-            value => Set("true", "false").contains(value)
+            values => !(values.contains("none") && values.size > 1)
```
Comment (Ewan-James-Donovan) [inline app/forms/AssessmentFormProvider.scala]: simply asserts that if 'none of the above' is selected, it is the only one selected. this enforced client-side, also.

File: app/controllers/AssessmentController.scala [hunk 3]
```diff
@@ -63,8 +63,8 @@ class AssessmentController @Inject() (
             .get(CategorisationDetailsQuery(recordId))
             .flatMap { categorisationInfo =>
               categorisationInfo.getAssessmentFromIndex(index).map { assessment =>
-                val listItems = assessment.getExemptionListItems
-                val form      = formProvider(listItems.size)
+                val codesAndDescriptions = assessment.getCodesZippedWithDescriptions
+                val form                 = formProvider()
 
                 val preparedForm = request.userAnswers.get(AssessmentPage(recordId, index)) match {
```
Comment (omotrium) [inline app/controllers/AssessmentController.scala]: we can use the generic prepared form method in the BaseController trait 

`val preparedForm = prepareForm(AssessmentPage(recordId, index), form)` 
you might need to merge the current main.

Comment (Ewan-James-Donovan) [reply app/controllers/AssessmentController.scala]: I'll merge main and update this - Done

File: app/forms/AssessmentFormProvider.scala (hunk 2 above)
Comment (omotrium) [reply app/forms/AssessmentFormProvider.scala]: can we refactor the form to this, just a suggestion let me know if you are okay.

`    Form(
      "value" -> seq(text(messagesKey))
        .verifying(messagesKey, values => values.nonEmpty)
        .verifying(messagesKey, values => values.size == 1 || !values.contains("none"))
        .transform[AssessmentAnswer](
          AssessmentAnswer.fromSeq,
          AssessmentAnswer.toSeq
        )
    )`

Comment (Ewan-James-Donovan) [reply app/forms/AssessmentFormProvider.scala]: Done! That's a nicer way of doing it :)

File: conf/messages.en [hunk 4]
```diff
@@ -271,19 +271,16 @@ traderReference.checkYourAnswersLabel = Trader reference
...
 assessment.title = Categorisation question
-assessment.heading = Category assessment {0}
-assessment.guidance = You need to confirm which of these exclusions you hold for {0}. We need this information to be able to categorise your goods.
+assessment.heading = Category assessment {0} of {1}
+assessment.guidance = You need to confirm which of these exclusions you hold for {0}. The theme for this assessment is:
 assessment.linkText = Search for a commodity code on the Online Tariff Tool (opens in new tab)
-assessment.question = Do you meet any of the following exclusions and can you supply the relevant document code on your movements?
-assessment.question.onlyOne = Do you meet the following exclusion and can you supply the relevant document code on your movement?
-assessment.hint = Select one answer.
+assessment.question = Which of these exclusions you hold for {0}?
```
Comment (besscerule) [inline conf/messages.en]: are we missing 'do' here?

File: conf/messages.en [hunk 5]
```diff
@@ -271,19 +271,16 @@ traderReference.checkYourAnswersLabel = Trader reference
...
+assessment.hint = Select all that apply.
 assessment.exemption = {0} - {1}
 assessment.exemption.none = None of the above
-assessment.error.required = Select yes if you meet any of the exclusions
-assessment.error.required.onlyOne = Select yes if you meet the exclusion
+assessment.error.required = Select exclusions you hold, or select ‘None of the above’
```
Comment (besscerule) [inline conf/messages.en]: Select THE exclusions you hold, or select ‘None of the above’

File: app/controllers/AssessmentController.scala [hunk 6]
```diff
@@ -63,24 +63,20 @@ class AssessmentController @Inject() (
             .get(CategorisationDetailsQuery(recordId))
             .flatMap { categorisationInfo =>
               categorisationInfo.getAssessmentFromIndex(index).map { assessment =>
-                val listItems = assessment.getExemptionListItems
```
Comment (besscerule) [inline app/controllers/AssessmentController.scala:66]: Not sure this is part of your ticket but the CYA stuff on this is broken so when you try to change something from the change link on the CYA page it doesn't persist

File: app/controllers/AssessmentController.scala [hunk 7]
```diff
@@ -63,24 +63,20 @@ class AssessmentController @Inject() (
...
-                val form      = formProvider(listItems.size)
```
Comment (besscerule) [inline app/controllers/AssessmentController.scala:67]: another CYA error is that I can't get it to show 'None of the above' on the CYA table - it just fills out the exemptions

File: app/controllers/AssessmentController.scala (hunk 6 above)
Comment (Ewan-James-Donovan) [reply app/controllers/AssessmentController.scala:66]: There's a linked ticket for CYA in the blocked section :)

File: app/controllers/AssessmentController.scala (hunk 7 above)
Comment (Ewan-James-Donovan) [reply app/controllers/AssessmentController.scala:67]: There's a linked ticket for CYA in the blocked section :)

File: conf/messages.en (hunk 4 above)
Comment (Ewan-James-Donovan) [reply conf/messages.en]: Yes we are :') Updated it

File: conf/messages.en (hunk 5 above)
Comment (Ewan-James-Donovan) [reply conf/messages.en]: Confluence said `Select exclusions you hold, or select ‘None of the above’` but I agree with you and have updated both.

**Review by besscerule — APPROVED** [review]

**Review by omotrium — APPROVED** [review]
lgtm

**Pull request merged**